### PR TITLE
[gatsby-source-graphcms] Add GraphCMS Plugin 

### PR DESCRIPTION
**Pull request by Redmega**
Adds the `gatsby-source-graphcms` plugin to fetch all data from a [GraphCMS](https://www.graphcms.com) backend.

Need to experiment with more complex data types and relationships to see if they're supported as is or if changes are needed.

As per @jorishermans in Discord:
> Let the user decide the mapping in the plugin config in gatsby-config.js
[...] I implemented a mapping feature in gatsby-source-mongodb
it would be nice if the mapping of mediatypes is almost the same in all the other plugins

## TODO: 
- [ ] Fix relationships/embedded fields
- [ ] Implement mapping feature for transformation plugins, like the MongoDB plugin
- [ ] Write tests
- [ ] Create demo site
> We can use the [Starwars API](https://api.graphcms.com/simple/v1/swapi) for tests and the demo site.

As per @mlukaszczyk from @GraphCMS, [Vinylbase](https://vinylbase.now.sh) is the "official" demo project for GraphCMS, and they'd be delighted to have a Gatsby clone of it using the same API (https://api.graphcms.com/simple/v1/vinylbase). Spike source for reference: https://github.com/GraphCMS/example_04_static_site_generation

#### This one pertains to the introspection metaquery method. I'll work on that later.
Example:

    Artist {
      picture {
        url
    ...
The metaQuery currently used is not capable of finding `url` from the above query, which will cause the query that fetches all data to fail. It will have to be modified. It looks like it will be a bit more difficult to find that `url` and add it to the query, as it might require making a `__type` query for each field of a Type that has subfields and then modifying the final query before firing it.

Errors can be seen here: https://github.com/Redmega/example_05_static_site_generation_with_gatsby

### 2. Embedded fields aren't found by GraphQL
Example: 

    query getAllArtists {
      allArtists {
        name
        id
        records {
          id
          title
          tracks {
            id
            title
            length
          }
        }
      }
    }
This query fed into the graphcms source plugin produces artists. name, id, records, id, and titles - but no tracks, or the subfields within. They exist in the redux store json, but graphql can't seem to identify them. More investigation is needed.

### CODE REVIEW

File: packages/gatsby-source-graphcms/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+Source plugin for pulling data into [Gatsby](https://github.com/gatsbyjs) from a GraphQL endpoint.
+
+## How to use
+*In your gatsby config...*
+```javascript
+plugins: [
+  /*
+   * Gatsby's data processing layer begins with “source”
+   * plugins. Here the site sources its data from a graphql endpoint
```
Comment (KyleAMathews) [inline packages/gatsby-source-graphcms/README.md]: "sources its data from the GraphCMS endpoint."

File: packages/gatsby-source-graphcms/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  {
+    resolve: `gatsby-source-graphcms`,
+    options: {
+      endpoint: `graphql_endpoint`,
+      token: `graphql_token`,
+    },
+  }
+],
+```
+I suggest either using a `.env` file or setting environment variables to access ala `process.env.GATSBY_GRAPHQL_ENDPOINT`, `process.env.GATSBY_GRAPHQL_TOKEN` so theres no endpoint and token on git ;)
```
Comment (KyleAMathews) [inline packages/gatsby-source-graphcms/README.md]: Linking to https://www.gatsbyjs.org/docs/environment-variables/ would be better.

That page needs updated to include a reference to using `process.env` for plugins.

Also you don't want to use "I" when writing docs as this is a community plugin so generally either just use command form "Use either a `.env` file or set environment variables directly…" or use "We". But "we" is rarely what you want.

File: packages/gatsby-source-graphcms/src/util.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import R from 'ramda'
+
+
+// If type ends in a non-vowel, we need to append es. Else s.
+export const formatTypeName = t => `all${t}${/s$/.test(t) ? `es` : `s`}`
```
Comment (KyleAMathews) [inline packages/gatsby-source-graphcms/src/util.js]: perhaps should use this instead? https://www.npmjs.com/package/pluralize

Or is this what GraphCMS says is their process?

Comment (Redmega) [inline packages/gatsby-source-graphcms/src/util.js]: I'll use `pluralize`. Forgot about all the special edge cases!

**Comment by Redmega** [comment]
We can use the [Starwars API](https://api.graphcms.com/simple/v1/swapi) for tests and the demo site.

File: packages/gatsby-source-graphcms/README.md (hunk 2 above)
Comment (Redmega) [inline packages/gatsby-source-graphcms/README.md]: Is there a special way to use process.env for plugins? Does the gatsby-config not have access to webpack's definitions at execution? I haven't tried it out yet, admittedly, so I will when I get home tonight.

File: packages/gatsby-source-graphcms/src/util.js (hunk 3 above)
Comment (KyleAMathews) [inline packages/gatsby-source-graphcms/src/util.js]: Well… what does GraphCMS do? That's what matters. Unless they use pluralize as well you'll just generate a bunch of invalid queries. Can't you get directly from the available types the connection name?

Comment (Redmega) [inline packages/gatsby-source-graphcms/src/util.js]: According to @GraphCMS
> [...] the process of pluralizing names is on their side. They use this library to do this: https://github.com/atteo/evo-inflector

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 4ded842ad181d48ea8cb2b2c97addd009191ddec

https://app.netlify.com/sites/using-glamor/deploys/599c9aa06f4c502da66fe33b

**Comment by Redmega** [comment]
For some reason when using the custom query to fetch data, the nodes are not being created properly. It is missing certain fields. I haven't had the chance to test it with different data, I'll do more tests tomorrow.

**Comment by Thebigbignooby** [comment]
looking forward to using this !

**Comment by Thebigbignooby** [comment]
what's the status on this ?

**Comment by Redmega** [comment]
@Thebigbignooby Been busy with work! Sorry!

Trying to get this out as simply as possible, but passing the query through directly causes some issues with the createNode function. It's not populating nested data more than one or two levels deep. I'll add more context to the main post tonight.

**Comment by rafacm** [comment]
I have created PR #2445 since I did not know how to add to this existing one.

/cc @Redmega @KyleAMathews @Thebigbignooby

**Comment by tlvenn** [comment]
Question, with the recent addition of schema stitching feature, could this be leveraged as a way to more easily plug a graphql endpoint as a data source in Gatsby ?

@KyleAMathews what do you think ?

**Comment by rafacm** [comment]
> with the recent addition of schema stitching feature, could this be leveraged 
> as a way to more easily plug a graphql endpoint as a data source in Gatsby ?

@tlvenn: could you elaborate?

**Comment by Redmega** [comment]
Schema stitching is a way to stitch schema from multiple sources together into a single query: apollographql/graphql-tools#382

**Comment by Redmega** [comment]
Continued in #2445
